### ISSUE TITLE
React Native support

### ISSUE THREAD

**Issue by magneticz** [mulesoft/js-client-oauth2#49]
Hi, first of all, thanks for the great work on library.

I am using this library on my react native project. It turns out, that production react-native app doesn't have some properties of window object, like window.btoa.

I imported base-64 library and imported it in source. What do you think of adding support for react-native? I could create PR

**Comment by blakeembrey** [comment]
I'd be happy to support React Native, it just has to work in a way that doesn't add any bloat to the runtime of the existing platforms (browser and node). Is there a feature flag or something else to switch behaviour in React Native like I do with `browser` in https://github.com/mulesoft/js-client-oauth2/blob/30fe581ac30f79d7445a3be5ef905099583d10e5/package.json#L13-L14? Did you have any issues doing HTTP requests also or does React Native follow the `browser` field and use `XMLHttpRequest` by default?

**Comment by magneticz** [comment]
I looked into it and it turns out there is a react-native field spec in package.json. 
I didn't experience any problems with HTTP requests. React Native provides Fetch API, that is why, I think, it is not a problem.

I created a branch, where i added support for btoa in react-native the way you did with default request. you can take a look here => https://github.com/magneticz/js-client-oauth2/tree/btoa-react-native-support

Let me know, if you think it needs to be done some other way. If it's good, i will be happy to make a PR

**Comment by blakeembrey** [comment]
That looks good to me. Would you mind pointing `/index.js` to the `react-native.js` version (instead of the folder)? In it's current state, and if it's working for you, I'll accept the PR and make a small adjustment to also split out the browser vs node `btoa`. Just make sure you also move the `btoaBuffer` function into https://github.com/magneticz/js-client-oauth2/blob/btoa-react-native-support/src/btoa/index.js.

**Comment by magneticz** [comment]
about adding /index.js, i spent 2 hours trying to figure out, why react-native flag is not working. It turns out, if I add index.js, it doesn't replace with platform specific file, so it seemed like the flag was not working. Long story short, if i add /index.js, react-native.js will not be substituted.

**Comment by magneticz** [comment]
also, querystring and url are not listed in the dependancies, are they available in other environments?
In react-native they are not available, so I had to add them as dependancies to my project, otherwise error is thrown.

**Comment by blakeembrey** [comment]
They are, yeah - Browserify normally fills in native node modules for you. I'm not sure what the preferred alternative there is, should they be listed as dependencies?

**Comment by magneticz** [comment]
I double checked, with index.js, react-native flag does not work, so it will not replace /btoa/index.js with /btoa/react-native.js. No idea why.
Also, I forgot to mention another error 
`TypeError: 'undefined' is not an object (evaluating 'window.location.protocol')`
In react native there is a window object, but window.location is undefined. I couldn't find where window.location.protocol is used as well. To fix it temporarily, i just wrote `window.location = {}` and it worked, not sure what is the right way, though. 

About those packages, i think they should be listed under dependancies, if this library has react-native support, because, for beginners it will not be clear, why they see that error (that querystring and url are not found)

**Comment by blakeembrey** [comment]
Do you know where `window.location` is coming from? The only globals I ever reference are `XMLHttpRequest` or `btoa`. Is it possibly from a separate package or something you installed (`url` or `querystring`)?

**Comment by magneticz** [comment]
it turns out `window.location.protocol` is used in `popsicle`.
```
if (window.location.protocol === 'https:' && /^http\:/.test(url)) {
    return reject(request.error("The request to \"" + url + "\" was blocked", 'EBLOCKED'));
}
```
Do you have any suggestions, on what could be done?

**Comment by blakeembrey** [comment]
That seems improbable since that's only exposed in the "browser" path of `popsicle` (not the regular node.js code) and if that's being followed it should also be followed here and not using `popsicle` anyway. Seems like you need to copy the section from `browser` into your new `react-native` section since the `react-native` section is replacing the previous `browser` configuration.

**Comment by magneticz** [comment]
Yeah, it worked, though I had to remove /index.js there as well. Created PR

**Comment by blakeembrey** [comment]
I'm really sorry I haven't been able to review it. I lost my access to this repo and sent an email in May, that seems to have been lost. FWIW, I'm working on a new authentication library over at https://github.com/blakeembrey/authwith which may be useful to you (or not). I'll be sure to add support for your use-case over there.

**Comment by blakeembrey** [comment]
Actually, which flow are you using on React Native - I could be wrong, I'm only supporting the single flow right now in `authwith` since it's designed for servers.

**Comment by magneticz** [comment]
i am not sure what you mean by flow?
and do you think you can merge PR i created for this library?

**Comment by blakeembrey** [comment]
I can not. I no longer have access to MuleSoft repositories, sorry. Hopefully someone else will be able to review it /cc @sichvoge

**Comment by andrroy** [comment]
@sichvoge Any update on this? It would be really nice to see this package working in react-native. Great work @magneticz !

**Comment by tarikd** [comment]
Hi guys, I have been playing with react native and spotify api, I stumbled upon client-oauth2 and I used it a few weeks ago. It worked flawlessly. I came back to my react native project today and I have an error every time I try to do a query on the api.

    btoa is not a function. (In 'btoa(toString(username) + ':' + toString(password))', 'btoa' is undefined)

Is this behavior normal, how did it work before and stoped working now? The only changes since I last ran the project is that I moved the folder and that there has been an update to Xcode.

**Comment by tarikd** [comment]
The error disapears when I activate the remote JS debugging in chrome.

**Comment by kenkku** [comment]
@tarikd When you use remote debugging in Chrome, the JS is run in Chrome, where `window.btoa` exists. In the native environment, it's not available.
